vhost-device-template/src/vhu_foo.rs

```diff
@@ -0,0 +1,409 @@
+                }
+            }
+
+            _ => {
+                warn!("unhandled device_event: {}", device_event);
+                return Err(Error::HandleEventUnknown.into());
+            }
+        }
+        Ok(())
+    }
+
+    fn exit_event(&self, _thread_index: usize) -> Option<EventFd> {
+        self.exit_event.try_clone().ok()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::mem::size_of;
+    use virtio_bindings::bindings::virtio_ring::{VRING_DESC_F_NEXT, VRING_DESC_F_WRITE};
+    use virtio_queue::{mock::MockSplitQueue, Descriptor, Queue};
+    use vm_memory::{Address, ByteValued, Bytes, GuestAddress, GuestMemoryAtomic, GuestMemoryMmap};
+
+    use super::*;
+
+    #[derive(Copy, Clone, Default)]
+    #[repr(C)]
+    struct VirtioFooOutHdr {
+        a: u16,
+        b: u16,
+        c: u32,
+    }
+    // SAFETY: The layout of the structure is fixed and can be initialized by
+    // reading its content from byte array.
+    unsafe impl ByteValued for VirtioFooOutHdr {}
+
+    #[derive(Copy, Clone, Default)]
+    #[repr(C)]
+    struct VirtioFooInHdr {
+        d: u8,
+    }
+    // SAFETY: The layout of the structure is fixed and can be initialized by
+    // reading its content from byte array.
+    unsafe impl ByteValued for VirtioFooInHdr {}
+
+    // Prepares a single chain of descriptors
+    fn prepare_desc_chain(start_addr: GuestAddress, buf: &mut Vec<u8>) -> FooDescriptorChain {
+        let mem = &GuestMemoryMmap::<()>::from_ranges(&[(start_addr, 0x1000)]).unwrap();
+        let vq = MockSplitQueue::new(mem, 16);
+        let mut next_addr = vq.desc_table().total_size() + 0x100;
+        let mut index = 0;
+
+        // Out header descriptor
+        let out_hdr = VirtioFooOutHdr {
+            a: 0x10,
+            b: 0x11,
+            c: 0x20,
+        };
+
+        let desc_out = Descriptor::new(
+            next_addr,
+            size_of::<VirtioFooOutHdr>() as u32,
+            VRING_DESC_F_NEXT as u16,
+            index + 1,
+        );
+
+        mem.write_obj::<VirtioFooOutHdr>(out_hdr, desc_out.addr())
+            .unwrap();
+        vq.desc_table().store(index, desc_out).unwrap();
+        next_addr += desc_out.len() as u64;
+        index += 1;
+
+        // Buf descriptor: optional
+        if !buf.is_empty() {
+            let desc_buf = Descriptor::new(
+                next_addr,
+                buf.len() as u32,
+                (VRING_DESC_F_WRITE | VRING_DESC_F_NEXT) as u16,
+                index + 1,
+            );
+            mem.write(buf, desc_buf.addr()).unwrap();
+            vq.desc_table().store(index, desc_buf).unwrap();
+            next_addr += desc_buf.len() as u64;
+            index += 1;
+        }
+
+        // In response descriptor
+        let desc_in = Descriptor::new(
+            next_addr,
+            size_of::<VirtioFooInHdr>() as u32,
+            VRING_DESC_F_WRITE as u16,
+            0,
+        );
+        vq.desc_table().store(index, desc_in).unwrap();
```

Comment:
This does the manual job of MockSplitQueue::build_desc_chain(), right? :thinking: Would it make sense to use that? I did something like that on vhu_scsi tests: https://github.com/rust-vmm/vhost-device/blob/31154ea0da7c7a716b388bccc3b08af58e9009bb/vhost-device-scsi/src/vhu_scsi.rs#L428-L442.
That seems a little less boilerplate rich. But I guess it can also be done when updating the other tests...

Reply:
> As you already suggest, can we do that in a separate pull request, since we need to fix existing implementations too ? Will you have time to do that ? Else I can take a look ..

I do not want to stop you if you want to fix it right now. Otherwise we can file a ticket and do it later. I won't get around it this week.

Reply:
> I do not want to stop you if you want to fix it right now.

I don't :)

Reply:
Sure! I will file a ticket to move this discussion off this PR.

Reply:
https://github.com/rust-vmm/vhost-device/issues/560

Reply:
Ah. I thought we take changing that out of the PR. But I can take a look now too!

Reply:
Naah, I changed my mind today and tried to fix it in this PR itself. Once it is sorted out, will fix other crates too.

Reply:
@vireshk: It does not work (cleanly). There is `build_chain`, but that is broken since it completely clears the flags (so it also clears the writable flag). `build_desc_chain` only works for one chain and then finalizes the queue. `build_multiple_desc_chains` assumes that descriptors already got chained together.

Some things are also a bit confusing `build_multiple_desc_chains` claims that it handles multiple chains. But then `build_desc_chain` calls it with what is supposed to be a single chain.

So I fear there is no quick fix. We will probably need to clean up the situation in the MockQueue and add something that can build multiple descriptor chains properly. It would probably also help to add some helper function that just takes a `Vec<Vec<(bool /* writable */, Vec<u8>)>>` and builds the entire vring + memory... I do not think there is need to make every device test jump through these mental gymnastics to get a simple test...

Hence, I cannot offer a quick fix :(.

Reply:
> Hence, I cannot offer a quick fix :(.

Exactly what I thought.